src/goto-programs/process_goto_program.cpp

```diff
@@ -55,6 +55,12 @@ bool process_goto_program(
   if(options.get_bool_option("partial-inline"))
   {
     log.status() << "Partial Inlining" << messaget::eom;
+    // Instrument preconditions can introduce locations that are not
+    // in correct order, while a postprocessing checking in the
+    // inlining can fail if locations are not in correct order.
+    // This refreshes the locations before inlining so any failures
+    // later are correctly attributed to errors in the inlining.
+    goto_model.goto_functions.compute_location_numbers();
```

Comment:
I think this needs to be in `instrument_preconditions` because "`update` was called after modification" should be an invariant of `goto_programt`'s and things that work on them.

In an ideal world we could add this to the `validate` checks and call them as post-condition of all of these transforms.